api/api-cache.php

```diff
@@ -0,0 +1,132 @@
+<?php
+// these functions use $db, so scripts referencing them need to define it
+
+/**
+ * Gets the first entry from the API cache that has a matching endpoint.
+ * @param string $endpoint The endpoint for the API request
+ * @return array|null The first entry that matches or null if no entry exists
+ */
+function getFromApiCacheTable(string $endpoint)
+{
+    global $db;
+    $q = $db->query('SELECT * FROM api_cache WHERE endpoint = "' . $db->real_escape_string($endpoint) . '"');
+    if ($q) {
+        $r = $q->fetch_array(MYSQLI_ASSOC);
+        return $r;
+    }
+    return null;
+}
+
+/**
+ * Gets the first entry from the API cache that has a matching endpoint. Will not return if the entry is expired.
+ * @param string $endpoint The endpoint for the API request
+ * @return array|false|null The first entry that matches, false if the entry is expired, or null if no entry exists
+ */
+function getCached(string $endpoint)
+{
+    global $db;
+    $r = getFromApiCacheTable($endpoint);
+    if ($r) {
+        $config = $db->query('SELECT * FROM outgoing_request_cache_config WHERE id = ' . $r['config_id'])->fetch_array(MYSQLI_ASSOC);
+        $ttl = $config['ttl'];
+        $now = time();
+        $retrievalTime = strtotime($r['last_successfully_retrieved'] . 'UTC');
+        if ($now - $retrievalTime <= $ttl) {
+            return $r;
+        }
+        return false;
+    }
+    return null;
+}
+
+/**
+ * Adds or updates an entry into the cache table.
+ * @param string $endpoint The endpoint for the API request
+ * @param string $content The content of API request's response
+ * @param int|string $config The ID of the config or its name in the config table
+ */
+function insertCached(string $endpoint, string $content, $config)
+{
+    global $db;
+    $r = getFromApiCacheTable($endpoint);
+    $queryString = "";
+    if ($r) {
+        $id = $r['id'];
+        $queryString = 'UPDATE api_cache 
+                        SET last_successfully_retrieved = UTC_TIMESTAMP, 
+                            last_attempted_retrieval = UTC_TIMESTAMP, 
+                            retry_count = 0, 
+                            content =\'' . $db->real_escape_string($content) . '\'
+                        WHERE id = ' . $id;
+
+    } else {
+        $configId = 1;
+        if (gettype($config) == 'integer') {
+            $configId = $config;
+        } else if (gettype($config) == 'string') {
+            $configId = $db->query('SELECT * FROM outgoing_request_cache_config WHERE config_name = "' . $db->real_escape_string($config) . '"')->fetch_array(MYSQLI_ASSOC)['id'];
+        }
+        $queryString = "INSERT INTO api_cache (endpoint, last_successfully_retrieved, last_attempted_retrieval, config_id, content) 
+                    values ('" . $db->real_escape_string($endpoint) . "', UTC_TIMESTAMP,
```

Comment:
I'm used to using `CURRENT_TIMESTAMP` for things like this, out of curiosity, is that just an alias for `UTC_TIMESTAMP` or is there actually a timezone difference between those two built in methods?

Reply:
`CURRENT_TIMESTAMP` returns the timestamp in the server's timezone. I was using UTC since I wasn't sure what the timezone on the server was, and I needed UTC since `time()` in PHP is a Unix timestamp (which is UTC).